client/src/providers/Authentication.js

```diff
@@ -0,0 +1,86 @@
+/* eslint-disable no-else-return */
+import React, { createContext, useReducer, useMemo, useContext, useEffect, useCallback } from "react";
+
+const initialState = { user: null, checked: false };
+const AuthenticationContext = createContext(initialState);
+
+const SIGN_IN = "authentication/signIn";
+const SIGN_OUT = "authentication/signOut";
+
+export const authenticationReducer = (state, action) => {
+  const { type, user } = action;
+  if (type === SIGN_IN) {
+    return { user, checked: true };
+  } else if (type === SIGN_OUT) {
+    return initialState;
+  } else {
+    return state;
+  }
+};
+
+export const authenticationActions = (state, dispatch) => {
+  return {
+    signIn: (user) => {
+      dispatch({
+        type: SIGN_IN,
+        user,
+      });
+    },
+    signOut: () => {
+      dispatch({
+        type: SIGN_OUT,
+      });
+    },
+  };
+};
+
+const AuthenticationProvider = ({ children }) => {
+  const [state, dispatch] = useReducer(authenticationReducer, initialState);
+  const actions = authenticationActions(state, dispatch);
+
+  const signOut = useCallback(() => fetch("/api/v1/user-sessions/logout")
+    .then((resp) => {
+      if (resp.ok && resp.status === 200) {
+        actions.signOut();
+      }
+    }), []);
+
+  const value = useMemo(
+    () => ({
+      ...state,
+      ...actions,
+      signOut
+    }),
+    [state, actions]
+  );
+
+  useEffect(() => {
```

Comment:
we should use the authentication or user client here as opposed to having native fetch calls here.